File: problems/spiralMatrix.js
// create a spiral matrix with N rows and N Columns, given N

function generate(N) {
  let res = Array.from(Array(4), () => []);
  let rowStart = 0,
    rowEnd = N - 1;
  let colStart = 0,
    colEnd = N - 1;

  let value = 1; // inserted val

  while (rowStart <= rowEnd && colStart <= colEnd) {
    // fill in first row
    for (let i = colStart; i <= colEnd; i++) {
      res[rowStart][i] = value;
      ++value;
    }
    rowStart++;

    // fill in RHS column
    for (let i = rowStart; i <= rowEnd; i++) {
      res[i][colEnd] = value;
      ++value;
    }
    colEnd--;

    // fill in bottom row
    for (let i = colEnd; i >= colStart; i--) {
      res[rowEnd][i] = value;
      ++value;
    }
    rowEnd--;

    // fill in lHS Col
    for (let i = rowEnd; i >= rowStart; i--) {
      res[i][colStart] = value;
      ++value;
    }
    colStart++;
  }

  return res;
}

console.log(generate(4));

// result
// [ [ 1, 2, 3, 4 ],
//   [ 12, 13, 14, 5 ],
//   [ 11, 16, 15, 6 ],
//   [ 10, 9, 8, 7 ] ]

File: problems/spiralMatrix.py


def generate(N):
    matrix = [[None]*N for num in range(N)]
    count = 1

    rowStart = 0
    rowEnd = N-1
    colStart = 0
    colEnd = N-1

    while(rowStart <= rowEnd and colStart <= colEnd):
        #     fill in top row:
        for col in range(colStart, colEnd+1):
            matrix[rowStart][col] = count
            count += 1
        rowStart += 1

        #  fill in right col
        for row in range(rowStart, rowEnd+1):
            matrix[row][colEnd] = count
            count += 1
        colEnd -= 1

        #  fill in the bottom row
        for col in range(colEnd, colStart-1, -1):
            matrix[rowEnd][col] = count
            count += 1
        rowEnd -= 1

        #  fill in the LHS col
        for row in range(rowEnd, rowStart-1, -1):
            matrix[row][colStart] = count
            count += 1
        colStart += 1

    print(matrix)


generate(4)
